sharedbuffers/mapped_struct.py

```diff
@@ -3869,6 +4049,447 @@ def merge(cls, parts, destfile = None, tempdir = None,
+                                if self._compare_keys(self._buf,
+                                         cython.cast(cython.p_uint, pindex + stride1)[0],
+                                         key):
+                                    return cython.cast(cython.p_uint, pindex + 2*stride1)[0]
+                                pindex += stride0
+                        finally:
+                            PyBuffer_Release(cython.address(indexbuf))
+                    else:
+                        index = self.index
+                        while startpos < nitems and index[startpos,0] == hkey:
+                            if self._compare_keys(self._buf, self.index[startpos,1], key):
+                                return index[startpos,2]
+                            startpos += 1
+                finally:
+                    PyBuffer_Release(cython.address(pybuf))
+                #lint:enable
+            else:
+                index = self.index
+                while startpos < nitems and index[startpos,0] == hkey:
+                    if self._compare_keys(buf, index[startpos, 1], key):
+                        return index[startpos,2]
+                    startpos += 1
+        return default
+
+    @classmethod
+    @cython.locals(
+        basepos = cython.ulonglong, curpos = cython.ulonglong, endpos = cython.ulonglong, finalpos = cython.ulonglong,
+        dtypemax = cython.ulonglong)
+    def build(cls, initializer, destfile = None, tempdir = None, map_file=True):
+        if destfile is None:
+            destfile = tempfile.NamedTemporaryFile(dir = tempdir)
+
+        dtype = cls.dtype
+        try:
+            dtypemax = ~dtype(0)
+        except:
+            try:
+                dtypemax = ~dtype.type(0)
+            except:
+                dtypemax = ~0
+
+        basepos = destfile.tell()
+
+        # Reserve space for the header
+        write = destfile.write
+        write(cls._Header.pack(0, 0))
+
+        # Build the index - the index is a matrix of the form:
+        # [ [ hash, key offset, value id ], ... ]
+        #
+        # With the rows ordered by hash
+        if isinstance(initializer, dict):
+            initializer = initializer.iteritems()
+        else:
+            initializer = iter(initializer)
+        parts = []
+        islice = itertools.islice
+        array = numpy.array
+        curpos = basepos + cls._Header.size
+        pack_into = mapped_object.pack_into
+        valbuf = bytearray(65536)
+        valbuflen = 65536
+        n = 0
+        part = []
+        while 1:
+            del part[:]
+            for k,i in islice(initializer, 1000):
+                # Add the index item
+                n += 1
+                # Not *quite* as correct as computing the actual length in
+                # bytes, but since this is only an upper bound, we should be fine.
+                klen = sys.getsizeof(k)
```

Comment:
Trouble is, you need the recursive size. This will probably fall short lots of times, and you're not retrying the way Schema.pack does.

You probably will have to implement the same retry mechanism schema.Pack uses, and you should also try to better estimate the size of the required buffer. You could have a configurable "default minimum buffer size" defined somewhere.

You can also reuse the bytearray for each object in the loop, you don't need a new byte array for each key. So you could just avoid trying to guess the right buffer size, and just start with a configurable buffer and expand the buffer when necessary (when pack_into raises an overflow exception, check Schema.pack).